<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Stephen Pederson" />


<title>Post-Alignment QC</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="site_libs/plotly-binding-4.10.0/plotly.js"></script>
<script src="site_libs/typedarray-0.1/typedarray.min.js"></script>
<link href="site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<link href="site_libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="site_libs/plotly-main-2.5.1/plotly-latest.min.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">20180328_Atkins_RatFracture</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="rawQC.html">Raw Data</a>
    </li>
    <li>
      <a href="trimQC.html">Trimmed Data</a>
    </li>
    <li>
      <a href="alnQC.html">Aligned Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="dge.html">Differential Expression</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Post-Alignment QC</h1>
<h4 class="author">Stephen Pederson</h4>
<h4 class="date">06 July, 2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-07-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>20180328_Atkins_RatFracture/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220705code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220705)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220705code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220705)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomUofABioinformaticsHub20180328AtkinsRatFracturetreedd288796c9511a31bec808d6c30df09e0f6d275ftargetblankdd28879a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/tree/dd288796c9511a31bec808d6c30df09e0f6d275f" target="_blank">dd28879</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomUofABioinformaticsHub20180328AtkinsRatFracturetreedd288796c9511a31bec808d6c30df09e0f6d275ftargetblankdd28879a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/tree/dd288796c9511a31bec808d6c30df09e0f6d275f" target="_blank">dd28879</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  data/gmt/
    Untracked:  output/dge.rds

Unstaged changes:
    Modified:   analysis/alnQC.Rmd
    Modified:   analysis/dge.Rmd
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/alnQC.Rmd</code>) and HTML
(<code>docs/alnQC.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/blob/dd288796c9511a31bec808d6c30df09e0f6d275f/analysis/alnQC.Rmd" target="_blank">dd28879</a>
</td>
<td>
Steve Pederson
</td>
<td>
2022-07-06
</td>
<td>
Setup initial DGE after restructure
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/dd288796c9511a31bec808d6c30df09e0f6d275f/docs/alnQC.html" target="_blank">dd28879</a>
</td>
<td>
Steve Pederson
</td>
<td>
2022-07-06
</td>
<td>
Setup initial DGE after restructure
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(ngsReports)
library(tidyverse)
library(yaml)
library(scales)
library(pander)
library(glue)
library(plotly)
library(edgeR)
library(ggfortify)
library(AnnotationHub)
library(ensembldb)
library(magrittr)
library(BSgenome.Rnorvegicus.UCSC.rn6)
library(broom)
library(ggrepel)</code></pre>
<pre class="r"><code>panderOptions(&quot;table.split.table&quot;, Inf)
panderOptions(&quot;big.mark&quot;, &quot;,&quot;)
theme_set(theme_bw())
suffix &lt;- &quot;_L001&quot;
pattern &lt;- paste0(&quot;_CB2YGANXX_.+fastq.gz&quot;)
sp &lt;- &quot;Rnorvegicus&quot;
rn6 &lt;- BSgenome.Rnorvegicus.UCSC.rn6</code></pre>
<pre class="r"><code>samples &lt;- &quot;data/targets.csv&quot; %&gt;%
  here::here() %&gt;%
  read_csv() %&gt;% 
    mutate(
      Filename = paste0(File, suffix)
    )</code></pre>
<pre class="r"><code>group_cols &lt;- hcl.colors(
  n = length(unique(samples$group)), 
  palette = &quot;Zissou 1&quot;
  ) %&gt;%
  setNames(unique(samples$group))</code></pre>
<div id="annotations" class="section level2">
<h2>Annotations</h2>
<div id="annotation-setup" class="section level3">
<h3>Annotation Setup</h3>
<pre class="r"><code>ah &lt;- AnnotationHub() %&gt;%
  subset(rdataclass == &quot;EnsDb&quot;) %&gt;%
  subset(species == &quot;Rattus norvegicus&quot;) %&gt;% 
  subset(str_detect(description, &quot;96&quot;))
ensDb &lt;- ah[[1]]
genesGR &lt;- genes(ensDb) %&gt;% 
  keepStandardChromosomes(species = &quot;Rattus_norvegicus&quot;, pruning.mode = &quot;coarse&quot;) %&gt;% 
  sortSeqlevels() 
transGR &lt;- transcripts(ensDb) %&gt;% 
  keepStandardChromosomes(species = &quot;Rattus_norvegicus&quot;, pruning.mode = &quot;coarse&quot;) %&gt;% 
  sortSeqlevels() 
exonGR &lt;- exonsBy(ensDb, &quot;tx&quot;) %&gt;% 
  reduce() %&gt;% 
  keepStandardChromosomes(species = &quot;Rattus_norvegicus&quot;, pruning.mode = &quot;coarse&quot;) %&gt;% 
  sortSeqlevels() 
exonGR_UCSC &lt;- exonGR
seqlevels(exonGR_UCSC) &lt;- paste0(&quot;chr&quot;, seqlevels(exonGR_UCSC)) %&gt;% 
  str_replace(&quot;chrMT&quot;, &quot;chrM&quot;)
genome(exonGR_UCSC) &lt;- genome(rn6)
exonSeq &lt;- getSeq(rn6, exonGR_UCSC) %&gt;% 
  lapply(unlist) %&gt;% 
  as(&quot;DNAStringSet&quot;)
transGC &lt;- exonSeq %&gt;% 
  letterFrequency(&quot;GC&quot;, as.prob = TRUE) %&gt;% 
  .[,1] %&gt;% 
  setNames(names(exonSeq))
transGR$gc_content &lt;- transGC[names(transGR)]
mcols(transGR) &lt;- mcols(transGR) %&gt;%
  cbind(
    transcriptLengths(ensDb)[rownames(.), c(&quot;nexon&quot;, &quot;tx_len&quot;)]
  )
mcols(genesGR) &lt;- mcols(genesGR) %&gt;%
  as.data.frame() %&gt;%
  dplyr::select(
    gene_id, gene_name, gene_biotype, entrezid
  ) %&gt;%
  left_join(
    mcols(transGR) %&gt;%
      as.data.frame() %&gt;%
      mutate(
        tx_support_level = case_when(
          is.na(tx_support_level) ~ 1L, 
          TRUE ~ tx_support_level
        )
      ) %&gt;%
      group_by(gene_id) %&gt;%
      summarise(
        n_tx = n(),
        longest_tx = max(tx_len),
        ave_tx_len = mean(tx_len),
        gc_content = sum(tx_len*gc_content) / sum(tx_len)
      ) %&gt;%
      mutate(
        bin_length = cut(
          x = ave_tx_len,
          labels = seq_len(10),
          breaks = quantile(ave_tx_len, probs = seq(0, 1, length.out = 11)),
          include.lowest = TRUE
        ),
        bin_gc = cut(
          x = gc_content,
          labels = seq_len(10),
          breaks = quantile(gc_content, probs = seq(0, 1, length.out = 11)),
          include.lowest = TRUE
        ),
        bin = paste(bin_gc, bin_length, sep = &quot;_&quot;)
      ),
    by = &quot;gene_id&quot;
  ) %&gt;%
  set_rownames(.$gene_id) %&gt;%
  as(&quot;DataFrame&quot;)
trans2Gene &lt;- mcols(transGR) %&gt;%
    as.data.frame() %&gt;%
    dplyr::select(tx_id, gene_id) %&gt;%
    dplyr::filter(!is.na(tx_id), !is.na(gene_id)) %&gt;%
    as_tibble()</code></pre>
<p>Annotation data was loaded as an <code>EnsDb</code> object, using
Ensembl release 96. Transcript level gene lengths and GC content was
converted to gene level values using:</p>
<ul>
<li><strong>GC Content</strong>: The total GC content divided by the
total length of transcripts</li>
<li><strong>Gene Length</strong>: The mean transcript length</li>
</ul>
<pre class="r"><code>write_rds(genesGR, here::here(&quot;output/genesGR.rds&quot;), compress = &quot;gz&quot;)</code></pre>
<pre class="r"><code>counts &lt;- list.files(here::here(&quot;data/3_kallisto&quot;), full.names = TRUE) %&gt;% 
    catchKallisto()
dge &lt;- counts$counts %&gt;%
    as.data.frame() %&gt;%
    rownames_to_column(&quot;tx_id&quot;) %&gt;%
    as_tibble() %&gt;%
    set_colnames(basename(colnames(.))) %&gt;%
    set_colnames(str_remove(colnames(.),&quot;_CB2Y.+&quot;)) %&gt;%
    mutate(tx_id = str_remove(tx_id, &quot;\\.[0-9]+&quot;)) %&gt;% 
  dplyr::filter(tx_id %in% trans2Gene$tx_id) %&gt;% 
  pivot_longer(cols = all_of(samples$Rat), names_to = &quot;Rat&quot;, values_to = &quot;count&quot;) %&gt;% 
    left_join(trans2Gene) %&gt;%
    group_by(Rat, gene_id) %&gt;%
    summarise(count = sum(count)) %&gt;% 
  pivot_wider(names_from = &quot;Rat&quot;, values_from = &quot;count&quot;) %&gt;% 
    dplyr::filter(grepl(&quot;ENSRNOG&quot;, gene_id)) %&gt;%
    as.data.frame() %&gt;% 
    column_to_rownames(&quot;gene_id&quot;) %&gt;% 
    DGEList() 
dge$samples %&lt;&gt;%
    mutate(Rat = rownames(.)) %&gt;%
  dplyr::select(-group) %&gt;% 
    left_join(samples, by = &quot;Rat&quot;) %&gt;%
    set_rownames(.$Rat)
dge$genes &lt;- genesGR[rownames(dge)] %&gt;% 
  mcols()</code></pre>
</div>
</div>
<div id="read-assignment-to-genes" class="section level1">
<h1>Read Assignment To Genes</h1>
<div id="library-sizes" class="section level2">
<h2>Library Sizes</h2>
<pre class="r"><code>dge$samples %&gt;% 
  mutate(lib.size = lib.size / 1e6) %&gt;% 
  ggplot(aes(Rat, lib.size, fill = group)) +
  geom_col() +
  facet_wrap(~group, scales = &quot;free_x&quot;) +
  scale_y_continuous(expand = expansion(c(0, .05))) +
  scale_fill_manual(values = group_cols) +
  labs(
    x = &quot;Sample&quot;, y = &quot;Library Size (millions)&quot;,
    fill = &quot;Group&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/alnQC.Rmd/plot-lib-sizes-1.png" alt="Library sizes after summarising to gene-level counts." width="960" />
<p class="caption">
Library sizes after summarising to gene-level counts.
</p>
</div>
<p>After assignment to genes, library sizes ranged between
<em>6,934,169</em> and <em>14,552,053</em> reads, with a median library
size of 10,208,651 reads.</p>
</div>
<div id="count-assignment-rates" class="section level2">
<h2>Count Assignment Rates</h2>
<pre class="r"><code>trimFqc &lt;- here::here(&quot;data/1_trimmedData/FastQC&quot;) %&gt;%
  list.files(pattern = &quot;zip&quot;, full.names = TRUE) %&gt;%
  FastqcDataList()
trimFqc %&gt;% 
  getModule(&quot;Basic&quot;) %&gt;% 
  mutate(Filename = str_remove_all(Filename, &quot;_R1.fastq.gz&quot;)) %&gt;% 
  left_join(dge$samples, by = &quot;Filename&quot;) %&gt;% 
  mutate(`% Assigned To Genes` = lib.size / Total_Sequences) %&gt;% 
  ggplot(aes(Rat, `% Assigned To Genes`, fill = group)) +
  geom_col() +
  facet_wrap(~group, scales = &quot;free_x&quot;) +
  scale_fill_manual(values = group_cols) +
  scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
  labs(fill = &quot;Group&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/alnQC.Rmd/plot-percent-assigned-1.png" alt="Assignment rate for reads to genes. All samples showed that a large proportion of reads were not derived from mRNA transcripts, and as such, were not informative for the analysis." width="960" />
<p class="caption">
Assignment rate for reads to genes. All samples showed that a large
proportion of reads were not derived from mRNA transcripts, and as such,
were not informative for the analysis.
</p>
</div>
</div>
<div id="total-detected-genes" class="section level2">
<h2>Total Detected Genes</h2>
<ul>
<li>Of the 23,706 genes defined in this annotation build, 1,621 genes
had no reads assigned in any samples.</li>
</ul>
<pre class="r"><code>dge$counts %&gt;%
  as_tibble() %&gt;%
  mutate(
    across(everything(), as.logical)
  ) %&gt;%
  summarise(
    across(everything(), sum)
  ) %&gt;%
  pivot_longer(
    everything(), names_to = &quot;Rat&quot;, values_to = &quot;Detected&quot;
  ) %&gt;%
  left_join(samples)%&gt;%
  ggplot(aes(group, Detected, colour = group)) +
  geom_point() +
  geom_segment(
    aes(xend = group, y = 0, yend = Detected),
    data = . %&gt;% 
      group_by(group) %&gt;%
      summarise(Detected = min(Detected)),
    colour = &quot;black&quot;, size = 1/4) +
  scale_y_continuous(labels = comma, expand = expansion(c(0, 0.05))) +
  scale_colour_manual(values = group_cols) +
  labs(
    x = &quot;Group&quot;, 
    y = &quot;Genes Detected&quot;,
    colour = &quot;Group&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/alnQC.Rmd/plotDetectedGenes-1.png" alt="*Total numbers of genes detected across all samples and groups.*" width="768" />
<p class="caption">
<em>Total numbers of genes detected across all samples and groups.</em>
</p>
</div>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotDetectedGenes-1">
Past versions of plotDetectedGenes-1.png
</button>
</p>
<div id="fig-plotDetectedGenes-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/blob/dd288796c9511a31bec808d6c30df09e0f6d275f/docs/figure/alnQC.Rmd/plotDetectedGenes-1.png" target="_blank">dd28879</a>
</td>
<td>
Steve Pederson
</td>
<td>
2022-07-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>plotly::ggplotly(
  dge$counts %&gt;%
    is_greater_than(0) %&gt;%
    rowSums() %&gt;%
    table() %&gt;%
    enframe(name = &quot;n_samples&quot;, value = &quot;n_genes&quot;) %&gt;%
    mutate(
      n_samples = as.integer(n_samples),
      n_genes = as.integer(n_genes),
    ) %&gt;%
    arrange(desc(n_samples)) %&gt;%
    mutate(
      Detectable = cumsum(n_genes),
      Undetectable = sum(n_genes) - Detectable
    ) %&gt;%
    pivot_longer(
      cols = ends_with(&quot;table&quot;),
      names_to = &quot;Status&quot;,
      values_to = &quot;Number of Genes&quot;
    ) %&gt;%
    dplyr::rename(
      `Number of Samples` = n_samples,
    ) %&gt;%
    ggplot(aes(`Number of Samples`, `Number of Genes`, colour = Status)) +
    geom_line() +
    geom_vline(
      aes(xintercept = `Mean Sample Number`),
      data = . %&gt;%
        summarise(`Mean Sample Number` = mean(`Number of Samples`)),
      linetype = 2,
      colour = &quot;grey50&quot;
    ) +
    scale_x_continuous(expand = expansion(c(0.01, 0.01))) +
    scale_y_continuous(labels = comma) +
    scale_colour_manual(values = c(rgb(0.1, 0.7, 0.2), rgb(0.7, 0.1, 0.1))) +
    labs(
      x = &quot;Samples &gt; 0&quot;
    )
)</code></pre>
<div class="figure" style="text-align: center">
<div id="htmlwidget-661d5ae1cbb6935a426f" style="width:768px;height:576px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-661d5ae1cbb6935a426f">{"x":{"data":[{"x":[0,1,2,3,4,5,6],"y":[23706,22085,20575,19230,18017,16605,15528],"text":["Number of Samples: 0<br />Number of Genes: 23706<br />Status: Detectable","Number of Samples: 1<br />Number of Genes: 22085<br />Status: Detectable","Number of Samples: 2<br />Number of Genes: 20575<br />Status: Detectable","Number of Samples: 3<br />Number of Genes: 19230<br />Status: Detectable","Number of Samples: 4<br />Number of Genes: 18017<br />Status: Detectable","Number of Samples: 5<br />Number of Genes: 16605<br />Status: Detectable","Number of Samples: 6<br />Number of Genes: 15528<br />Status: Detectable"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(26,179,51,1)","dash":"solid"},"hoveron":"points","name":"Detectable","legendgroup":"Detectable","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,1,2,3,4,5,6],"y":[0,1621,3131,4476,5689,7101,8178],"text":["Number of Samples: 0<br />Number of Genes:     0<br />Status: Undetectable","Number of Samples: 1<br />Number of Genes:  1621<br />Status: Undetectable","Number of Samples: 2<br />Number of Genes:  3131<br />Status: Undetectable","Number of Samples: 3<br />Number of Genes:  4476<br />Status: Undetectable","Number of Samples: 4<br />Number of Genes:  5689<br />Status: Undetectable","Number of Samples: 5<br />Number of Genes:  7101<br />Status: Undetectable","Number of Samples: 6<br />Number of Genes:  8178<br />Status: Undetectable"],"type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(179,26,26,1)","dash":"solid"},"hoveron":"points","name":"Undetectable","legendgroup":"Undetectable","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[3,3],"y":[-1185.3,24891.3],"text":"Mean Sample Number: 3","type":"scatter","mode":"lines","line":{"width":1.88976377952756,"color":"rgba(127,127,127,1)","dash":"dash"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":25.7412480974125,"r":7.30593607305936,"b":39.6955859969559,"l":60.6392694063927},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-0.06,6.06],"tickmode":"array","ticktext":["0","2","4","6"],"tickvals":[6.93889390390723e-18,2,4,6],"categoryorder":"array","categoryarray":["0","2","4","6"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"Samples > 0","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1185.3,24891.3],"tickmode":"array","ticktext":["0","5,000","10,000","15,000","20,000"],"tickvals":[0,5000,10000,15000,20000],"categoryorder":"array","categoryarray":["0","5,000","10,000","15,000","20,000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"Number of Genes","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"title":{"text":"Status","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"7b85a4acabe":{"x":{},"y":{},"colour":{},"type":"scatter"},"7b85399b7092":{"xintercept":{}}},"cur_data":"7b85a4acabe","visdat":{"7b85a4acabe":["function (y) ","x"],"7b85399b7092":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<p class="caption">
<em>Total numbers of genes detected shown against the number of samples
with at least one read assigned to each gene.</em>
</p>
</div>
</div>
<div id="pca" class="section level2">
<h2>PCA</h2>
<div id="sample-similarity" class="section level3">
<h3>Sample Similarity</h3>
<pre class="r"><code>pca &lt;- dge$counts %&gt;%
  .[rowSums(. == 0) &lt; ncol(.)/2,] %&gt;%
  cpm(log = TRUE) %&gt;%
  t() %&gt;%
  prcomp()</code></pre>
<p>A PCA was performed using logCPM values from the subset of 18,017
genes with at least one read in more than half of the samples.</p>
<pre class="r"><code>showLabel &lt;- nrow(samples) &lt;= 20
pca %&gt;% 
tidy() %&gt;% 
  dplyr::rename(Rat = row) %&gt;% 
  left_join(dge$samples, by = &quot;Rat&quot;) %&gt;% 
  dplyr::filter(PC %in% 1:2) %&gt;% 
  pivot_wider(names_from = &quot;PC&quot;, names_prefix = &quot;PC&quot;, values_from = &quot;value&quot;) %&gt;% 
  ggplot(
    aes(PC1, PC2, colour = group, size = lib.size/1e6)
  ) +
  geom_point() +
  geom_text_repel(aes(label = Rat), show.legend = FALSE) +
  scale_colour_manual(values = group_cols) +
  scale_size_continuous(limits = c(5, 15), breaks = seq(5, 15, by = 5)) +
  labs(
    x = glue(&quot;PC1 ({percent(pca$sdev[[1]]^2 / sum(pca$sdev^2), 0.1)})&quot;),
    y = glue(&quot;PC2 ({percent(pca$sdev[[2]]^2 / sum(pca$sdev^2), 0.1)})&quot;),
    colour = &quot;Group&quot;,
    size = &quot;Library Size\n(millions)&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="figure/alnQC.Rmd/plotPCA-1.png" alt="*PCA plot of all samples. PC1 is most strongly correlated with library size, whilst PC2 appears to capture the majority of the biological variability*" width="768" />
<p class="caption">
<em>PCA plot of all samples. PC1 is most strongly correlated with
library size, whilst PC2 appears to capture the majority of the
biological variability</em>
</p>
</div>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotPCA-1">
Past versions of plotPCA-1.png
</button>
</p>
<div id="fig-plotPCA-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/blob/dd288796c9511a31bec808d6c30df09e0f6d275f/docs/figure/alnQC.Rmd/plotPCA-1.png" target="_blank">dd28879</a>
</td>
<td>
Steve Pederson
</td>
<td>
2022-07-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="gc-and-length-biases" class="section level3">
<h3>GC and Length Biases</h3>
<pre class="r"><code>mcols(genesGR) %&gt;%
  as.data.frame() %&gt;%
  dplyr::filter(gene_id %in% rownames(pca$rotation)) %&gt;%
  as_tibble() %&gt;%
  mutate(
    bin_length = cut(
      x = ave_tx_len,
      labels = seq_len(10),
      breaks = quantile(ave_tx_len, probs = seq(0, 1, length.out = 11)),
      include.lowest = TRUE
    ),
    bin_gc = cut(
      x = gc_content,
      labels = seq_len(10),
      breaks = quantile(gc_content, probs = seq(0, 1, length.out = 11)),
      include.lowest = TRUE
    ),
    bin = paste(bin_gc, bin_length, sep = &quot;_&quot;)
  ) %&gt;%
  dplyr::select(gene_id, contains(&quot;bin&quot;)) %&gt;%
  mutate(
    PC1 = pca$rotation[gene_id, &quot;PC1&quot;],
    PC2 = pca$rotation[gene_id, &quot;PC2&quot;]
  ) %&gt;%
  pivot_longer(
    cols = c(&quot;PC1&quot;, &quot;PC2&quot;),
    names_to = &quot;PC&quot;,
    values_to = &quot;value&quot;
  ) %&gt;%
  group_by(PC, bin_gc, bin_length, bin) %&gt;%
  summarise(
    Size = n(),
    mean = mean(value),
    sd = sd(value),
    t = t.test(value)$statistic,
    p = t.test(value)$p.value,
    adjP = p.adjust(p, method = &quot;bonf&quot;)
  ) %&gt;%
  ggplot(
    aes(bin_length, bin_gc, colour = t, alpha = -log10(adjP), size = Size)
  ) +
  geom_point() +
  facet_wrap(~PC) +
  scale_colour_gradient2() +
  scale_size_continuous(range = c(1, 10)) +
  labs(
    x = &quot;Average Transcript Length&quot;,
    y = &quot;GC Content&quot;,
    alpha = expression(paste(-log[10], p[adj]))) +
  theme(
    panel.grid = element_blank(),
    legend.position = &quot;bottom&quot;
    ) </code></pre>
<div class="figure" style="text-align: center">
<img src="figure/alnQC.Rmd/plotPCA_GC_Length-1.png" alt="*Contribution of each GC/Length Bin to PC1 and PC2. Fill colours indicate the t-statistic, with tranparency denoting significance as -log10(p), using Bonferroni-adjusted p-values.*" width="960" />
<p class="caption">
<em>Contribution of each GC/Length Bin to PC1 and PC2. Fill colours
indicate the t-statistic, with tranparency denoting significance as
-log10(p), using Bonferroni-adjusted p-values.</em>
</p>
</div>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plotPCA_GC_Length-1">
Past versions of plotPCA_GC_Length-1.png
</button>
</p>
<div id="fig-plotPCA_GC_Length-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/UofABioinformaticsHub/20180328_Atkins_RatFracture/blob/dd288796c9511a31bec808d6c30df09e0f6d275f/docs/figure/alnQC.Rmd/plotPCA_GC_Length-1.png" target="_blank">dd28879</a>
</td>
<td>
Steve Pederson
</td>
<td>
2022-07-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>write_rds(dge, here::here(&quot;output/dge.rds&quot;))</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggrepel_0.9.1                       broom_0.8.0                        
 [3] BSgenome.Rnorvegicus.UCSC.rn6_1.4.1 BSgenome_1.64.0                    
 [5] rtracklayer_1.56.0                  Biostrings_2.64.0                  
 [7] XVector_0.36.0                      magrittr_2.0.3                     
 [9] ensembldb_2.20.1                    AnnotationFilter_1.20.0            
[11] GenomicFeatures_1.48.0              AnnotationDbi_1.58.0               
[13] Biobase_2.56.0                      GenomicRanges_1.48.0               
[15] GenomeInfoDb_1.32.1                 IRanges_2.30.0                     
[17] S4Vectors_0.34.0                    AnnotationHub_3.4.0                
[19] BiocFileCache_2.4.0                 dbplyr_2.1.1                       
[21] ggfortify_0.4.14                    edgeR_3.38.0                       
[23] limma_3.52.0                        plotly_4.10.0                      
[25] glue_1.6.2                          pander_0.6.5                       
[27] scales_1.2.0                        yaml_2.3.5                         
[29] forcats_0.5.1                       stringr_1.4.0                      
[31] dplyr_1.0.9                         purrr_0.3.4                        
[33] readr_2.1.2                         tidyr_1.2.0                        
[35] tidyverse_1.3.1                     ngsReports_1.13.0                  
[37] tibble_3.1.7                        ggplot2_3.3.6                      
[39] BiocGenerics_0.42.0                 workflowr_1.7.0                    

loaded via a namespace (and not attached):
  [1] readxl_1.4.0                  backports_1.4.1              
  [3] lazyeval_0.2.2                crosstalk_1.2.0              
  [5] BiocParallel_1.30.0           digest_0.6.29                
  [7] htmltools_0.5.2               fansi_1.0.3                  
  [9] memoise_2.0.1                 tzdb_0.3.0                   
 [11] modelr_0.1.8                  matrixStats_0.62.0           
 [13] vroom_1.5.7                   prettyunits_1.1.1            
 [15] colorspace_2.0-3              blob_1.2.3                   
 [17] rvest_1.0.2                   rappdirs_0.3.3               
 [19] haven_2.5.0                   xfun_0.30                    
 [21] callr_3.7.0                   crayon_1.5.1                 
 [23] RCurl_1.98-1.6                jsonlite_1.8.0               
 [25] zoo_1.8-10                    gtable_0.3.0                 
 [27] zlibbioc_1.42.0               DelayedArray_0.22.0          
 [29] Rhdf5lib_1.18.0               DBI_1.1.2                    
 [31] Rcpp_1.0.8.3                  viridisLite_0.4.0            
 [33] xtable_1.8-4                  progress_1.2.2               
 [35] bit_4.0.4                     DT_0.22                      
 [37] htmlwidgets_1.5.4             httr_1.4.3                   
 [39] ellipsis_0.3.2                farver_2.1.0                 
 [41] pkgconfig_2.0.3               XML_3.99-0.9                 
 [43] sass_0.4.1                    here_1.0.1                   
 [45] locfit_1.5-9.5                utf8_1.2.2                   
 [47] labeling_0.4.2                tidyselect_1.1.2             
 [49] rlang_1.0.2                   later_1.3.0                  
 [51] munsell_0.5.0                 BiocVersion_3.15.2           
 [53] cellranger_1.1.0              tools_4.2.0                  
 [55] cachem_1.0.6                  cli_3.3.0                    
 [57] generics_0.1.2                RSQLite_2.2.13               
 [59] evaluate_0.15                 fastmap_1.1.0                
 [61] ggdendro_0.1.23               processx_3.5.3               
 [63] knitr_1.39                    bit64_4.0.5                  
 [65] fs_1.5.2                      KEGGREST_1.36.0              
 [67] whisker_0.4                   mime_0.12                    
 [69] xml2_1.3.3                    biomaRt_2.52.0               
 [71] compiler_4.2.0                rstudioapi_0.13              
 [73] filelock_1.0.2                curl_4.3.2                   
 [75] png_0.1-7                     interactiveDisplayBase_1.34.0
 [77] reprex_2.0.1                  bslib_0.3.1                  
 [79] stringi_1.7.6                 highr_0.9                    
 [81] ps_1.7.0                      lattice_0.20-45              
 [83] ProtGenerics_1.28.0           Matrix_1.4-1                 
 [85] vctrs_0.4.1                   rhdf5filters_1.8.0           
 [87] pillar_1.7.0                  lifecycle_1.0.1              
 [89] BiocManager_1.30.17           jquerylib_0.1.4              
 [91] data.table_1.14.2             bitops_1.0-7                 
 [93] httpuv_1.6.5                  R6_2.5.1                     
 [95] BiocIO_1.6.0                  promises_1.2.0.1             
 [97] gridExtra_2.3                 MASS_7.3-57                  
 [99] assertthat_0.2.1              rhdf5_2.40.0                 
[101] SummarizedExperiment_1.26.1   rprojroot_2.0.3              
[103] rjson_0.2.21                  withr_2.5.0                  
[105] GenomicAlignments_1.32.0      Rsamtools_2.12.0             
[107] GenomeInfoDbData_1.2.8        parallel_4.2.0               
[109] hms_1.1.1                     grid_4.2.0                   
[111] rmarkdown_2.14                MatrixGenerics_1.8.0         
[113] git2r_0.30.1                  getPass_0.2-2                
[115] shiny_1.7.1                   lubridate_1.8.0              
[117] restfulr_0.0.13              </code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
