---
title: "Post-Alignment QC"
author: "Stephen Pederson"
date: "`r format(Sys.Date(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.align = "center",
  fig.height = 7,
  fig.width = 10
)
```

```{r packages}
library(ngsReports)
library(tidyverse)
library(yaml)
library(scales)
library(pander)
library(glue)
library(plotly)
library(edgeR)
library(ggfortify)
library(AnnotationHub)
library(ensembldb)
library(magrittr)
library(BSgenome.Rnorvegicus.UCSC.rn6)
library(broom)
library(ggrepel)
```

```{r options}
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
theme_set(theme_bw())
suffix <- "_L001"
pattern <- paste0("_CB2YGANXX_.+fastq.gz")
sp <- "Rnorvegicus"
rn6 <- BSgenome.Rnorvegicus.UCSC.rn6
```


```{r samples}
samples <- "data/targets.csv" %>%
  here::here() %>%
  read_csv() %>% 
	mutate(
	  Filename = paste0(File, suffix)
	)
```


```{r group_cols}
group_cols <- hcl.colors(
  n = length(unique(samples$group)), 
  palette = "Zissou 1"
  ) %>%
  setNames(unique(samples$group))
```

## Annotations

### Annotation Setup

```{r ah}
ah <- AnnotationHub() %>%
  subset(rdataclass == "EnsDb") %>%
  subset(species == "Rattus norvegicus") %>% 
  subset(str_detect(description, "96"))
ensDb <- ah[[1]]
genesGR <- genes(ensDb) %>% 
  keepStandardChromosomes(species = "Rattus_norvegicus", pruning.mode = "coarse") %>% 
  sortSeqlevels() 
transGR <- transcripts(ensDb) %>% 
  keepStandardChromosomes(species = "Rattus_norvegicus", pruning.mode = "coarse") %>% 
  sortSeqlevels() 
exonGR <- exonsBy(ensDb, "tx") %>% 
  reduce() %>% 
  keepStandardChromosomes(species = "Rattus_norvegicus", pruning.mode = "coarse") %>% 
  sortSeqlevels() 
exonGR_UCSC <- exonGR
seqlevels(exonGR_UCSC) <- paste0("chr", seqlevels(exonGR_UCSC)) %>% 
  str_replace("chrMT", "chrM")
genome(exonGR_UCSC) <- genome(rn6)
exonSeq <- getSeq(rn6, exonGR_UCSC) %>% 
  lapply(unlist) %>% 
  as("DNAStringSet")
transGC <- exonSeq %>% 
  letterFrequency("GC", as.prob = TRUE) %>% 
  .[,1] %>% 
  setNames(names(exonSeq))
transGR$gc_content <- transGC[names(transGR)]
mcols(transGR) <- mcols(transGR) %>%
  cbind(
    transcriptLengths(ensDb)[rownames(.), c("nexon", "tx_len")]
  )
mcols(genesGR) <- mcols(genesGR) %>%
  as.data.frame() %>%
  dplyr::select(
    gene_id, gene_name, gene_biotype, entrezid
  ) %>%
  left_join(
    mcols(transGR) %>%
      as.data.frame() %>%
      mutate(
        tx_support_level = case_when(
          is.na(tx_support_level) ~ 1L, 
          TRUE ~ tx_support_level
        )
      ) %>%
      group_by(gene_id) %>%
      summarise(
        n_tx = n(),
        longest_tx = max(tx_len),
        ave_tx_len = mean(tx_len),
        gc_content = sum(tx_len*gc_content) / sum(tx_len)
      ) %>%
      mutate(
        bin_length = cut(
          x = ave_tx_len,
          labels = seq_len(10),
          breaks = quantile(ave_tx_len, probs = seq(0, 1, length.out = 11)),
          include.lowest = TRUE
        ),
        bin_gc = cut(
          x = gc_content,
          labels = seq_len(10),
          breaks = quantile(gc_content, probs = seq(0, 1, length.out = 11)),
          include.lowest = TRUE
        ),
        bin = paste(bin_gc, bin_length, sep = "_")
      ),
    by = "gene_id"
  ) %>%
  set_rownames(.$gene_id) %>%
  as("DataFrame")
trans2Gene <- mcols(transGR) %>%
	as.data.frame() %>%
	dplyr::select(tx_id, gene_id) %>%
	dplyr::filter(!is.na(tx_id), !is.na(gene_id)) %>%
	as_tibble()
```

Annotation data was loaded as an `EnsDb` object, using Ensembl release 96.
Transcript level gene lengths and GC content was converted to gene level values using:

- **GC Content**: The total GC content divided by the total length of transcripts
- **Gene Length**: The mean transcript length

```{r exportGenesGR}
write_rds(genesGR, here::here("output/genesGR.rds"), compress = "gz")
```


```{r counts, results='hide'}
counts <- list.files(here::here("data/3_kallisto"), full.names = TRUE) %>% 
	catchKallisto()
dge <- counts$counts %>%
	as.data.frame() %>%
	rownames_to_column("tx_id") %>%
	as_tibble() %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.),"_CB2Y.+")) %>%
	mutate(tx_id = str_remove(tx_id, "\\.[0-9]+")) %>% 
  dplyr::filter(tx_id %in% trans2Gene$tx_id) %>% 
  pivot_longer(cols = all_of(samples$Rat), names_to = "Rat", values_to = "count") %>% 
	left_join(trans2Gene) %>%
	group_by(Rat, gene_id) %>%
	summarise(count = sum(count)) %>% 
  pivot_wider(names_from = "Rat", values_from = "count") %>% 
	dplyr::filter(grepl("ENSRNOG", gene_id)) %>%
	as.data.frame() %>% 
	column_to_rownames("gene_id") %>% 
	DGEList() 
dge$samples %<>%
	mutate(Rat = rownames(.)) %>%
  dplyr::select(-group) %>% 
	left_join(samples, by = "Rat") %>%
	set_rownames(.$Rat)
dge$genes <- genesGR[rownames(dge)] %>% 
  mcols()
```



# Read Assignment To Genes

## Library Sizes

```{r plot-lib-sizes, fig.cap = "Library sizes after summarising to gene-level counts."}
dge$samples %>% 
  mutate(lib.size = lib.size / 1e6) %>% 
  ggplot(aes(Rat, lib.size, fill = group)) +
  geom_col() +
  facet_wrap(~group, scales = "free_x") +
  scale_y_continuous(expand = expansion(c(0, .05))) +
  scale_fill_manual(values = group_cols) +
  labs(
    x = "Sample", y = "Library Size (millions)",
    fill = "Group"
  )
```

After assignment to genes, library sizes ranged between `r pander(comma(range(dge$samples$lib.size)))` reads, with a median library size of `r pander(comma(median(dge$samples$lib.size)))` reads.

## Count Assignment Rates

```{r plot-percent-assigned, fig.cap = "Assignment rate for reads to genes. All samples showed that a large proportion of reads were not derived from mRNA transcripts, and as such, were not informative for the analysis."}
trimFqc <- here::here("data/1_trimmedData/FastQC") %>%
  list.files(pattern = "zip", full.names = TRUE) %>%
  FastqcDataList()
trimFqc %>% 
  getModule("Basic") %>% 
  mutate(Filename = str_remove_all(Filename, "_R1.fastq.gz")) %>% 
  left_join(dge$samples, by = "Filename") %>% 
  mutate(`% Assigned To Genes` = lib.size / Total_Sequences) %>% 
  ggplot(aes(Rat, `% Assigned To Genes`, fill = group)) +
  geom_col() +
  facet_wrap(~group, scales = "free_x") +
  scale_fill_manual(values = group_cols) +
  scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
  labs(fill = "Group")
```


## Total Detected Genes

- Of the `r comma(nrow(dge))` genes defined in this annotation build, `r comma(sum(rowSums(dge$counts) == 0))` genes had no reads assigned in any samples.

```{r plotDetectedGenes, fig.height=6, fig.width=8, fig.cap = "*Total numbers of genes detected across all samples and groups.*"}
dge$counts %>%
  as_tibble() %>%
  mutate(
    across(everything(), as.logical)
  ) %>%
  summarise(
    across(everything(), sum)
  ) %>%
  pivot_longer(
    everything(), names_to = "Rat", values_to = "Detected"
  ) %>%
  left_join(samples)%>%
  ggplot(aes(group, Detected, colour = group)) +
  geom_point() +
  geom_segment(
    aes(xend = group, y = 0, yend = Detected),
    data = . %>% 
      group_by(group) %>%
      summarise(Detected = min(Detected)),
    colour = "black", size = 1/4) +
  scale_y_continuous(labels = comma, expand = expansion(c(0, 0.05))) +
  scale_colour_manual(values = group_cols) +
  labs(
    x = "Group", 
    y = "Genes Detected",
    colour = "Group"
  )
```

```{r plotDetectedVsSamples, fig.height=6, fig.width=8, fig.cap = "*Total numbers of genes detected shown against the number of samples with at least one read assigned to each gene.*"}
plotly::ggplotly(
  dge$counts %>%
    is_greater_than(0) %>%
    rowSums() %>%
    table() %>%
    enframe(name = "n_samples", value = "n_genes") %>%
    mutate(
      n_samples = as.integer(n_samples),
      n_genes = as.integer(n_genes),
    ) %>%
    arrange(desc(n_samples)) %>%
    mutate(
      Detectable = cumsum(n_genes),
      Undetectable = sum(n_genes) - Detectable
    ) %>%
    pivot_longer(
      cols = ends_with("table"),
      names_to = "Status",
      values_to = "Number of Genes"
    ) %>%
    dplyr::rename(
      `Number of Samples` = n_samples,
    ) %>%
    ggplot(aes(`Number of Samples`, `Number of Genes`, colour = Status)) +
    geom_line() +
    geom_vline(
      aes(xintercept = `Mean Sample Number`),
      data = . %>%
        summarise(`Mean Sample Number` = mean(`Number of Samples`)),
      linetype = 2,
      colour = "grey50"
    ) +
    scale_x_continuous(expand = expansion(c(0.01, 0.01))) +
    scale_y_continuous(labels = comma) +
    scale_colour_manual(values = c(rgb(0.1, 0.7, 0.2), rgb(0.7, 0.1, 0.1))) +
    labs(
      x = "Samples > 0"
    )
)
```



## PCA

### Sample Similarity

```{r prcomp}
pca <- dge$counts %>%
  .[rowSums(. == 0) < ncol(.)/2,] %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

A PCA was performed using logCPM values from the subset of `r comma(sum(rowSums(dge$counts == 0) < ncol(dge)/2))` genes with at least one read in more than half of the samples.

```{r plotPCA, fig.height=6, fig.width=8, fig.cap="*PCA plot of all samples. PC1 is most strongly correlated with library size, whilst PC2 appears to capture the majority of the biological variability*"}
showLabel <- nrow(samples) <= 20
pca %>% 
tidy() %>% 
  dplyr::rename(Rat = row) %>% 
  left_join(dge$samples, by = "Rat") %>% 
  dplyr::filter(PC %in% 1:2) %>% 
  pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>% 
  ggplot(
    aes(PC1, PC2, colour = group, size = lib.size/1e6)
  ) +
  geom_point() +
  geom_text_repel(aes(label = Rat), show.legend = FALSE) +
  scale_colour_manual(values = group_cols) +
  scale_size_continuous(limits = c(5, 15), breaks = seq(5, 15, by = 5)) +
  labs(
    x = glue("PC1 ({percent(pca$sdev[[1]]^2 / sum(pca$sdev^2), 0.1)})"),
    y = glue("PC2 ({percent(pca$sdev[[2]]^2 / sum(pca$sdev^2), 0.1)})"),
    colour = "Group",
    size = "Library Size\n(millions)"
  )
```

### GC and Length Biases

```{r plotPCA_GC_Length, fig.height=6, fig.cap = "*Contribution of each GC/Length Bin to PC1 and PC2. Fill colours indicate the t-statistic, with tranparency denoting significance as -log10(p), using Bonferroni-adjusted p-values.*"}
mcols(genesGR) %>%
  as.data.frame() %>%
  dplyr::filter(gene_id %in% rownames(pca$rotation)) %>%
  as_tibble() %>%
  mutate(
    bin_length = cut(
      x = ave_tx_len,
      labels = seq_len(10),
      breaks = quantile(ave_tx_len, probs = seq(0, 1, length.out = 11)),
      include.lowest = TRUE
    ),
    bin_gc = cut(
      x = gc_content,
      labels = seq_len(10),
      breaks = quantile(gc_content, probs = seq(0, 1, length.out = 11)),
      include.lowest = TRUE
    ),
    bin = paste(bin_gc, bin_length, sep = "_")
  ) %>%
  dplyr::select(gene_id, contains("bin")) %>%
  mutate(
    PC1 = pca$rotation[gene_id, "PC1"],
    PC2 = pca$rotation[gene_id, "PC2"]
  ) %>%
  pivot_longer(
    cols = c("PC1", "PC2"),
    names_to = "PC",
    values_to = "value"
  ) %>%
  group_by(PC, bin_gc, bin_length, bin) %>%
  summarise(
    Size = n(),
    mean = mean(value),
    sd = sd(value),
    t = t.test(value)$statistic,
    p = t.test(value)$p.value,
    adjP = p.adjust(p, method = "bonf")
  ) %>%
  ggplot(
    aes(bin_length, bin_gc, colour = t, alpha = -log10(adjP), size = Size)
  ) +
  geom_point() +
  facet_wrap(~PC) +
  scale_colour_gradient2() +
  scale_size_continuous(range = c(1, 10)) +
  labs(
    x = "Average Transcript Length",
    y = "GC Content",
    alpha = expression(paste(-log[10], p[adj]))) +
  theme(
    panel.grid = element_blank(),
    legend.position = "bottom"
    ) 
```

```{r export}
write_rds(dge, here::here("output/dge.rds"))
```

