---
title: "Differential Gene Expression"
author: "Stephen Pederson"
date: "`r format(Sys.Date(), '%d %B, %Y')`"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.align = "center",
  fig.height = 7,
  fig.width = 10
)
```

```{r packages}
library(tidyverse)
library(yaml)
library(scales)
library(pander)
library(glue)
library(edgeR)
library(AnnotationHub)
library(ensembldb)
library(magrittr)
library(cqn)
library(cowplot)
library(matrixStats)
library(broom)
library(ggrepel)
library(statmod)
```

```{r options}
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
theme_set(theme_bw())
suffix <- "_L001"
pattern <- paste0("_CB2YGANXX_.+fastq.gz")
sp <- "Rnorvegicus"
```


```{r samples}
samples <- "data/targets.csv" %>%
  here::here() %>%
  read_csv() %>% 
	mutate(
	  Filename = paste0(File, suffix)
	)
```


```{r group_cols}
group_cols <- hcl.colors(
  n = length(unique(samples$group)), 
  palette = "Zissou 1"
  ) %>%
  setNames(unique(samples$group))
```

# Setup

## Annotations

```{r ah}
ah <- AnnotationHub() %>%
  subset(rdataclass == "EnsDb") %>%
  subset(species == "Rattus norvegicus") %>% 
  subset(str_detect(description, "96"))
ensDb <- ah[[1]]
genesGR <- read_rds(here::here("output/genesGR.rds"))
transGR <- transcripts(ensDb) %>% 
  subset(gene_id %in% names(genesGR))
```

Gene annotations were again loaded from Ensembl Release `r ensemblVersion(ensDb)`.
The [previously defined](alnQC.html#Annotation_Setup) `GenomicRanges` object containing GC content and Gene Length was also loaded,
containing information for `r comma(length(genesGR))` genes.

## Count Data

```{r counts}
# counts <- here::here("data/2_alignedData/featureCounts/counts.out") %>%
#   read_tsv(comment = "#") %>%
#   rename_all(str_remove_all, pattern = "_CB2Y.++") %>%
#   rename_all(basename) %>%
#   dplyr::select(-Chr, -Start, -End, -Strand, -Length) %>%
#   column_to_rownames("Geneid") %>%
#   as.matrix()
counts <- list.files(here::here("data/3_kallisto"), full.names = TRUE) %>% 
	catchKallisto()
```

```{r}
trans2Gene <- mcols(transGR) %>%
	as.data.frame() %>%
	dplyr::select(tx_id, gene_id) %>%
	dplyr::filter(!is.na(tx_id), !is.na(gene_id)) %>%
	as_tibble()
```

```{r geneDGE}
dge <- counts$counts %>%
	as.data.frame() %>%
	rownames_to_column("tx_id") %>%
	as_tibble() %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.),"_CB2Y.+")) %>%
	mutate(tx_id = str_remove(tx_id, "\\.[0-9]+")) %>% 
  dplyr::filter(tx_id %in% trans2Gene$tx_id) %>% 
  pivot_longer(cols = all_of(samples$Rat), names_to = "Rat", values_to = "count") %>% 
	left_join(trans2Gene) %>%
	group_by(Rat, gene_id) %>%
	summarise(count = sum(count)) %>% 
  pivot_wider(names_from = "Rat", values_from = "count") %>% 
	dplyr::filter(grepl("ENSRNOG", gene_id)) %>%
	as.data.frame() %>% 
	column_to_rownames("gene_id") %>% 
	.[rowSums(cpm(.) > 2) >= 3,] %>%
	DGEList() 
dge$samples %<>%
	mutate(Rat = rownames(.)) %>%
  dplyr::select(-group) %>% 
	left_join(samples, by = "Rat") %>%
	set_rownames(.$Rat)
dge$genes <- genesGR[rownames(dge)] %>% 
  mcols()
```

```{r dge_featre-counts, eval = FALSE}
dge <- DGEList(
  counts = counts,
  samples = column_to_rownames(samples, "Rat")[colnames(counts),],
  genes = mcols(genesGR[rownames(counts)])[, c("gene_id", "gene_name", "gene_biotype", "entrezid", "longest_tx", "ave_tx_len", "gc_content")]
) %>%
  calcNormFactors()
dge$samples$Rat <- rownames(dge$samples)
```

Prior to filtering for undetectable genes, counts were loaded as a `DGEList`, incorporating both sample and gene metadata.

# Normalisation

```{r genes2keep}
genes2Keep <- cpm(dge) %>% 
  is_greater_than(1) %>% 
  rowSums() %>% 
  is_weakly_greater_than(3)
```


## Before Normalisation {.tabset}

Genes were defined as detected if exceeding 1 CPM in at least 3 samples, leading to
`r comma(sum(genes2Keep))` genes being retained as confidently expressed prior to normalisation.

### Density Plots

```{r plot-density-all, fig.cap = "Distributions of logCPM values for all non-zero genes, and those retained as detected using the above CPM threshold. The number of genes in each category is also shown."}
list(
  All = cpm(dge, log = TRUE) %>% 
    as.data.frame() %>% 
    mutate(which = "All Genes"),
  Detected = cpm(dge, log = TRUE)[genes2Keep, ] %>% 
    as.data.frame() %>% 
    mutate(which = "Detected Genes")
) %>% 
  bind_rows() %>% 
  pivot_longer(
    cols = all_of(samples$Rat), names_to = "Rat", values_to = "logCPM"
  )%>% 
  left_join(samples, by = "Rat") %>% 
  ggplot(
    aes(
      logCPM, colour = group, group = Rat
    )        
  ) +
  geom_density() +
  facet_wrap(~which) +
  geom_text(
    aes(x, y, label = lab),
    data = . %>% 
      group_by(which) %>% 
      summarise(
        x = 0.85*max(logCPM),
        y = 0.85*max(density(logCPM)$y),
        lab = glue("n = {comma(n() / ncol(dge))}"),
        .groups = "drop"
      ) %>% 
      mutate(y = max(y)),
    inherit.aes = FALSE
  )+
  scale_colour_manual(values = group_cols) +
  labs(
    y = "Density", colour = "Group"
  )
```

### RLE

```{r plot-rle-pre, fig.cap = "Relative Log Expression. Any deviations from zero indicate potential batch effects."}
cpm(dge, log = TRUE) %>% 
  # .[genes2Keep, ] %>%
  as.data.frame() %>% 
  pivot_longer(
    cols = everything(), names_to = "Rat", values_to = "logCPM"
  ) %>% 
  left_join(samples) %>% 
  group_by(Rat) %>% 
  mutate(RLE = logCPM - median(logCPM))%>% 
  ggplot(aes(Rat, RLE, fill = group)) +
  geom_boxplot(alpha = 0.9) +
  geom_hline(yintercept = 0, linetype = 2) +
  facet_wrap(~group, scales = "free_x") +
  scale_fill_manual(values = group_cols) +
  labs(
    x = "Sample", fill = "Group"
  )
```

## After Normalisation {.tabset}

```{r cqn}
cqn <- dge %>% 
  with(
    suppressWarnings(
      cqn(
        counts = counts,
        x = genes$gc_content,
        lengths = genes$ave_tx_len
      )
    )
  )
dgeNorm <- dge
dgeNorm$offset <- cqn$offset
logCPM <- cqn$y + cqn$offset
```

### CQN Diagnostic Plots

```{r plot-cqn, fig.cap = 'Diagnostic plots from CQN. Any divergence between samples indicates a range of either GC content or gene length where some bias was observed.'}
a <- cqn$func1 %>%
  as.data.frame() %>%
  mutate(x = cqn$grid1) %>%
  pivot_longer(
    cols = any_of(colnames(dge)),
    names_to = "Rat",
    values_to = "QR fit"
  ) %>%
  left_join(dge$samples, by = "Rat") %>%
  ggplot(
    aes(x, `QR fit`, group = Rat, colour = group)
  ) +
  geom_line() +
  scale_colour_manual(values = group_cols) +
  labs(
    x = "GC content", colour = "Group"
  )
b <- cqn$func2 %>%
  as.data.frame() %>%
  mutate(x = cqn$grid2) %>%
  pivot_longer(
    cols = any_of(colnames(dge)),
    names_to = "Rat",
    values_to = "QR fit"
  ) %>%
  left_join(dge$samples, by = "Rat") %>%
  ggplot(
    aes(x, `QR fit`, colour = group, group = Rat
    )
  ) +
  geom_line() +
  scale_colour_manual(values= group_cols) +
  labs(
    x = expression(paste(log[10], " Gene Length (kb)")),  colour = "Group"
  )
plot_grid(
  a + theme(legend.position = "none"), 
  b + theme(legend.position = "none"),
  get_legend(a),
  labels = c("A", "B"),
  nrow = 1,
  rel_widths = c(3, 3, 1)
)
```

### Density

```{r plot-density-norm, fig.cap = "logCPM densities before and after Conditional-Quantile Normalisation."}
list(
  pre = cpm(dgeNorm$counts, log = TRUE) %>% 
    as.data.frame() %>% 
    mutate(status = "Raw") ,
  post = logCPM %>% 
    as.data.frame() %>% 
    mutate(status = "CQ-Normalised")
) %>% 
  bind_rows() %>% 
  mutate(status = fct_rev(status)) %>% 
  pivot_longer(
    cols = all_of(samples$Rat), names_to = "Rat", values_to = "logCPM"
  ) %>% 
  left_join(samples, by = "Rat") %>% 
  ggplot(
    aes(logCPM, colour = group, group = Rat)       
  ) +
  geom_density() +
  facet_wrap(~status) +
  scale_colour_manual(values = group_cols) +
  labs(
    y = "Density", colour = "Group"
  )
```


### PCA

```{r plot-pca-post, fig.cap = "PCA on normalised logCPM values"}
pcaPost <- logCPM %>%
  .[rowVars(.) > 0,] %>% 
  t() %>%
  prcomp() 
pcaPost %>%
  tidy() %>% 
  dplyr::rename(Rat = row) %>% 
  left_join(samples, by = "Rat") %>% 
  dplyr::filter(PC %in% 1:2) %>% 
  pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>% 
  ggplot(
    aes(PC1, PC2, colour = group)
  ) +
  geom_point() +
  geom_text_repel(aes(label = Rat), show.legend = FALSE) +
  scale_colour_manual(values = group_cols) +
  labs(
    x = glue("PC1 ({percent(pcaPost$sdev[[1]]^2 / sum(pcaPost$sdev^2), 0.1)})"),
    y = glue("PC2 ({percent(pcaPost$sdev[[2]]^2 / sum(pcaPost$sdev^2), 0.1)})"),
    colour = "Group"
  )
```

# Analysis

```{r glmfit}
X <- model.matrix(~group, dgeNorm$samples) %>% 
  set_colnames(str_remove(colnames(.), "group"))
dgeNorm <- estimateDisp(dgeNorm, design = X, robust = TRUE)
fit <- glmQLFit(dgeNorm, design = X)
topTable <- glmQLFTest(fit, coef = "Diabetic") %>%
  topTags(n = Inf) %>%
  .[["table"]] %>%
  as_tibble() %>%
  mutate(
    rankingStat = -sign(logFC)*log10(PValue),
    signedRank = rank(rankingStat),
    DE = FDR < 0.05
  )
```

```{r ma-plot}
topTable %>%
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = DE),alpha = 0.5) +
  geom_text_repel(
    aes(label = gene_name, colour = DE),
    data = . %>%
      arrange(desc(abs(logFC))) %>% 
      dplyr::slice(1:20),
    show.legend = FALSE
  ) +
  geom_smooth(se = FALSE) +
  scale_colour_manual(values = c("grey50", "red")) +
  theme(legend.position = "none")
```

```{r volcano-plot}
topTable %>%
  ggplot(aes(logFC, -log10(PValue))) +
  geom_point(aes(colour = DE),alpha = 0.5) +
  geom_text_repel(
    aes(label = gene_name, colour = DE),
    data = . %>%
      arrange(desc(abs(logFC))) %>% 
      dplyr::slice(1:20),
    show.legend = FALSE
  ) +
  scale_colour_manual(values = c("grey50", "red")) +
  theme(legend.position = "none")
```
