---
title: "QC On Raw Data"
date: "`r format(Sys.Date(), '%d %B, %Y')`"
---

```{r setup}
knitr::opts_chunk$set(
	message = FALSE, warning = FALSE,
	fig.height = 6, fig.width = 10
)
```

```{r packages}
library(ngsReports)
library(tidyverse)
library(yaml)
library(scales)
library(pander)
library(glue)
library(plotly)
```

```{r options}
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
theme_set(theme_bw())
suffix <- "_L001_R1.fastq.gz"
pattern <- paste0("_CB2YGANXX_.+fastq.gz")
sp <- "Rnorvegicus"
```


```{r samples}
samples <- "data/targets.csv" %>%
  here::here() %>%
  read_csv() %>% 
	mutate(
	  Filename = paste0(File, suffix)
	)
```


```{r group_cols}
group_cols <- hcl.colors(
  n = length(unique(samples$group)), 
  palette = "Zissou 1"
  ) %>%
  setNames(unique(samples$group))
```


# Quality Assessment on Raw Data

```{r rawFqc}
rawFqc <- here::here("data/0_rawData/FastQC") %>%
  list.files(pattern = "zip", full.names = TRUE) %>%
  FastqcDataList() %>%
  .[fqName(.) %in% samples$Filename]
```

```{r plotSummary, fig.height=fh, fig.cap = "*Overall summary of FastQC reports*"}
plotSummary(rawFqc, pattern = pattern)
```

## Library Sizes

A total of `r length(rawFqc)` libraries were contained in this dataset, with read totals ranging between `r pander(comma(range(readTotals(rawFqc)$Total_Sequences)))` reads.

Across all libraries, reads were `r unique(getModule(rawFqc, "Sequence_Length")$Length)` bases.

```{r plotReadTotals, fig.cap = "*Library Sizes for all supplied fastq files. Any samples run as multiple libraries are  shown as the supplied multiple libraries and have not been merged.*"}
plotReadTotals(rawFqc, pattern = pattern, usePlotly = TRUE)
```

## Sequence Quality

```{r plotBaseQuals, fig.cap = "*Mean sequencing quality scores at each base position for each library*"}
plotBaseQuals(
  rawFqc,
  pattern = pattern, 
  usePlotly = TRUE,
  dendrogram = TRUE,
  cluster = TRUE
  )
```


## GC Content

```{r plotGC, fig.cap=glue("*GC content shown as the % above and below the theoretical GC content for the {sp} transcriptome. The peaks at 62% and 71% are strongly suggestive of incomplete rRNA removal*")}
plotGcContent(
  x = rawFqc, 
  pattern = pattern, 
  species = sp, 
  plotType = "line",
  gcType = "Trans",
  usePlotly = TRUE,
  dendrogram = TRUE,
  cluster = TRUE
  )
```

## Sequence Content

```{r plotSeqContentResid, fig.cap = "*Base and Position specific residuals for each sample. The mean base content at each position was calculated for each nucleotide, and the sample-specific residuals calculated.*"}
plotly::ggplotly(
  getModule(rawFqc, module = "Per_base_sequence_content") %>% 
    mutate(Base = fct_inorder(Base)) %>%
    group_by(Base) %>% 
    mutate(
      across(c("A", "C", "G", "T"), function(x){x - mean(x)}) 
    ) %>% 
    pivot_longer(
      cols = c("A", "C", "G", "T"), 
      names_to = "Nuc", 
      values_to = "resid"
    ) %>%
    left_join(samples) %>%
    ggplot(
      aes(Base, resid, group = Filename, colour = group)
    ) + 
    geom_line() +
    facet_wrap(~Nuc) + 
    scale_colour_manual(values = group_cols) +
    labs(
      x = "Read Position", y = "Residual", colour = "Group"
    ) +
    theme(
      axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
    )
)
```

## AdapterContent

```{r plotAdapterContent, fig.cap="*Total Adapter Content for each sample shown by starting position in the read.*"}
plotAdapterContent(
  x = rawFqc, 
  pattern = pattern, 
  plotType = "line",
  dendrogram = TRUE,
  cluster = TRUE
  ) +
  scale_y_continuous(expand = expansion(c(0, 0)), limits = c(0, 10))
```

## Overrepresented Sequences

```{r os}
os <- suppressMessages(getModule(rawFqc, "Over"))
```

```{r plotOverRep, fig.cap = "*Summary of over-represented sequences across all libraries*"}
if (nrow(os)){
  if (length(unique(os$Filename)) > 1){
    suppressMessages(
      plotOverrep(
        x = rawFqc,
        pattern = suffix, 
        usePlotly = TRUE,
        dendrogram = TRUE,
        cluster = TRUE
      )
    )
  }
}
```

```{r}
os %>%
  group_by(Sequence, Possible_Source) %>%
  summarise(
    `Found in` = n(),
    Total = sum(Count),
    `Largest Percent` = glue("{round(max(Percentage), 2)}%")
  ) %>%
  pander(
    caption = "*Summary of over-represented sequences within the raw data. Manual checking of over-represented sequences using BLAST indicated most were Ribosomal in origin*"
  )
```

<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
```{r session-info, echo=FALSE}
pander::pander(sessionInfo())
```
</div>