---
title: "Diabetic Fracture: Differential Gene Analysis"
author: "Steve Pederson"
date: "22 May 2019"
output: 
  html_document: 
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
					  warning = FALSE, message = FALSE,
					  fig.align = "center")
```

## Setup

```{r packages}
library(limma)
library(edgeR)
library(tidyverse)
library(magrittr)
library(pander)
library(ngsReports)
library(scales)
library(GenomicRanges)
library(rtracklayer)
library(plyranges)
library(ggrepel)
library(fgsea)
library(parallel)
theme_set(theme_bw())
```

```{r rawFqc}
rawFqc <- list.files("../0_rawData/FastQC", pattern = "zip", full.names = TRUE) %>%
	getFastqcData()
```


## Data Description & Quality Assessment

- The initial files were single-end reads of 100bp
- Libraries appeared to be stranded and contained Total RNA-like properties, not polyA
- Library sizes were between `r pander(comma(range(readTotals(rawFqc)$Total_Sequences)))` reads, which was considered as good
- Low levels of adapter content were found and removed using `AdapterRemoval`. This is normal and occurs when a small number of fragments < 100bp are sequenced
- Spikes in GC content at around 63% and 72% were noted as potentially indicative of incomplete rRNA removal
    - This raised the question as to whether human rRNA sequences targeted rather than rat during rRNA depletion?
    - rRNA-aligning reads were *not* removed from the fastq files as they will be ignored during summarisation at the gene-level. They will however be expected to align multiple times when aligning to the reference genome
- Reads were aligned to the Rnor6.0 genome sourced from Ensembl Release 96, using `STAR 2.7.0d`
    + These alignments were only used for visualisation and inspection with regard to QC
- For Differential Gene Expression (DGE) analysis, reads were pseudo-aligned to the transcriptome from Ensembl Release 96 for summarisation at the transcript-level using `kallisto 0.43.1`. 
    + `kallisto` is able to proportionally assign multi-mapping reads across both genes and transcripts
    +  This approach only counts *reads which align to spliced transcripts*, as contained in the reference transcriptome, with the correct strandedness. Intronic reads are not counted, and neither are reads which align to the opposite strand from the transcript  
    + These transcript-level counts were used for DGE analysis after being summarised to the gene-level. There was not sufficient read depth to attempt transcript-level analysis.


```{r plotGC, fig.cap="*GC content of each library showing uneven and 'spikey' patterns. Peaks on the RHS indicate potential incomplete rRNA removal, whilst the other peaks indicate possible excessive PCR cycles and high levels of duplicated sequences.*",fig.width=8}
plotGcContent(rawFqc, plotType = "line", theoreticalGC = TRUE, theoreticalType = "Transcriptome", species = "Rnorvegicus", pattern = "_CB2.+")
```

```{r plotDup, fig.cap = "*Estimated duplication levels within each sample. Higher than expected sequence duplication rates were observed, particularly for sequences present greater than 10,000 times. Sample 179_F and 187_F were of particular concern, however this is also suggestive of incomplete rRNA removal*"}
Sequence_Duplication_Levels(rawFqc) %>%
	mutate(Duplication_Level = factor(Duplication_Level, levels = unique(Duplication_Level)),
		   Filename = str_extract(Filename, "[0-9]+_F")) %>%
	ggplot(aes(Duplication_Level, Percentage_of_total, colour = Filename)) +
	geom_line(aes(group = Filename)) +
	labs(x = "Duplication Level",
		 y = "Percentage of Total Library")
```


```{r starLogs}
starLogs <- list.files(
	path = "../2_alignedData/log/",
	pattern = "final.out",
	full.names = TRUE) %>%
	importStarLogs()
```

```{r plotAlignments, fig.cap = "*STAR Alignment summary showing very low levels of unique alignments, which is again suggestive of incomplete rRNA removal. Uniquely mapping reads are within the range usually seen after summarising to genes, indicating that final gene-level counts may be a little lower than is optimal.*", fig.width=8}
starLogs %>% 
	dplyr::select(Filename, contains("Number")) %>% 
	dplyr::select(Filename, contains("Mapped")) %>% 
	gather("Type", "Total", -Filename) %>% 
	mutate(Filename = str_remove(Filename, "_CB2Y.+"),
		   Type = str_remove(Type, "Number_Of_Reads_"),
		   Type = str_remove(Type, "_Reads_Number")) %>% 
	ggplot(aes(Filename, Total, fill = Type)) + 
	geom_bar(stat = "identity") +
	scale_y_continuous(label = comma) +
	labs(x = "Rat ID",
		 y = "Reads")
```

- Manual inspection of the alignments also indicated that the number of PCR rounds during library preparation may have been slightly more than required.
- PCR duplicates are difficult to identify with 100% confidence in paired-end libraries, and are not able to be removed confidently at all in single-end libraries
- Gene expression estimates in the presence of a high duplication rate are quite stable for moderate to highly expressed genes, however expression estimates for low-expressed genes are less reliable
- As a result, low-expressed genes will be removed to avoid false positives.

```{r counts, results='hide'}
counts <- list.files("../3_kallisto/", full.names = TRUE) %>% 
	catchKallisto()
```

```{r targets}
targets <- read_csv("targets.csv")
```


After assignment to transcripts, the library sizes ranged between `r pander(comma(range(colSums(counts$counts))))` reads

- This is a disappointing result, but still manageable
- These library sizes were lower than might have been expected after inspection of the STAR alignment rate. Inspection of STAR alignments revealed that a significant number of reads were not derived from a spliced transcript, or were aligned to the incorrect strand


## Initial Analysis

```{r setupGRanges, cache=TRUE}
allGR <- import.gff("ftp://ftp.ensembl.org/pub/release-96/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.96.chr.gtf.gz")
geneGR <- allGR %>%
	subset(type == "gene") %>%
	plyranges::select(type, gene_id, gene_name, gene_biotype)
# transGR
transGR <- allGR %>%
	subset(type == "transcript") %>%
	plyranges::select(gene_id, gene_name, transcript_id, transcript_name, transcript_biotype)
```

```{r trans2Gene}
trans2Gene <- mcols(transGR) %>%
	as.data.frame() %>%
	dplyr::select(transcript_id, gene_id) %>%
	dplyr::filter(!is.na(transcript_id), !is.na(gene_id)) %>%
	as_tibble()
```

```{r geneDGE}
geneDGE <- counts$counts %>%
	as.data.frame() %>%
	rownames_to_column("transcript_id") %>%
	as_tibble() %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.),"_CB2Y.+")) %>%
	mutate(transcript_id = str_remove(transcript_id, "\\.[0-9]+")) %>%
	gather(Sample, Count, -transcript_id) %>%
	left_join(trans2Gene) %>%
	group_by(Sample, gene_id) %>%
	summarise(Count = sum(Count)) %>%
	spread(Sample, Count) %>% 
	dplyr::filter(grepl("ENSRNOG", gene_id)) %>%
	as.data.frame() %>% 
	column_to_rownames("gene_id") %>% 
	.[rowSums(cpm(.) > 2) >= 3,] %>%
	DGEList() %>%
	calcNormFactors()
geneDGE$samples %<>%
	mutate(Rat = rownames(.)) %>%
	left_join(targets) %>%
	set_rownames(.$Rat)
geneDGE$genes <- geneGR %>%
	magrittr::set_names(mcols(.)$gene_id) %>%
	.[rownames(geneDGE)]
```

```{r voomData}
voomData <- voomWithQualityWeights(geneDGE) 
```


```{r pca}
pca <- geneDGE %>%
	cpm(log = TRUE) %>%
	t() %>%
	prcomp()
```


- The first step of the analysis was to summarise transcript-level counts to the gene-level
- Data was analysed using [voom-precision weights](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29), which allows for the assumption of normality despite counts being a *discrete* variable.
    + This also allows for a weighted regression analysis to be performed in which lower-confidence samples can be down-weighted. Standard t-tests for differential expression can also be employed, with moderated variances to reduce false positives & false negatives
    + Sample weights were calculated by assuming that all samples were replicates, which produces more conservative weights. Given that the vast majority of genes are not differentially-expressed, this is also a realistic assumption.
- A PCA was performed using counts transformed to log~2~ CPM (counts per million) which also contains a normalisation step.
    + The first principal component captured `r percent(summary(pca)$importance["Proportion of Variance","PC1"])` of the variance
    + The first two components captured a total of `r percent(summary(pca)$importance["Cumulative Proportion","PC2"])`, confirming a strong signal in the data
    + PC1 tracked strongly with library size which was slightly concerning
    + The difference between treatment groups **was clearly evident on PC2** which confirmed a strong biological signal within the experiment, despite PC1.


```{r plotPCA, fig.cap = "*PCA with colours indicating treatment group and point sizes indicating sample weights. A trend from left to right along PC1 tracked somewhat with library size, with all points along PC1 appearing in order of library size for each respective treatment group. Point sizes are proportional to sample weights.*"}
plotly::ggplotly(
	pca$x %>%
		cbind(voomData$targets[rownames(.),]) %>%
		mutate(lib.size = comma(lib.size)) %>%
		ggplot(aes(PC1, PC2, colour = Group, label = Rat, lib = lib.size)) +
		geom_point(aes(size = sample.weights)) +
		labs(x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance", "PC1"]), ")"),
			 y = paste0("PC2 (", percent(summary(pca)$importance["Proportion of Variance", "PC2"]), ")")) +
		theme(legend.position = "none")
)
```

```{r PC1vsLibSize, fig.cap="*Correlation between library size and PC1. A simple regression line is also shown for illustrative purposes*"}
pca$x %>%
	cbind(voomData$targets[rownames(.),]) %>%
	ggplot(aes(PC1, lib.size, label = Rat)) +
	geom_point(aes(size = sample.weights, colour = Group)) +
	geom_text_repel() +
	geom_label(x = -60, y = 1.5e7, label = expression(rho == 0.794)) +
	geom_smooth(method = "lm", se = FALSE) +
	scale_y_continuous(label = comma) +
	theme(legend.position = "none")
```

### Differentially Expressed Genes

Differentially expressed (DE) genes were detected using a slight variation from the common null hypothesis.
Instead of testing $H_0: \log FC = 0$ which tests for the point value zero, the null hypothesis was defined as being within the range [$-\tau, \tau$], with $\tau$ set at the default value of $\log_2 1.2$.
This additionally saves filtering any detected DE based on fold-change as is common practice in many workflows.

$$
H_0: |\log FC| \leq \tau  \\
vs  \\
H_A: |\log FC| > \tau
$$

```{r results}
results <- voomData %>%
	lmFit(design = model.matrix(~Group, data = .$targets)) %>%
	# eBayes() %>%
	# topTable(n = Inf) %>%
	treat() %>%
	topTreat(n = Inf, coef = "GroupDiabetic") %>%
	set_colnames(str_remove(colnames(.),"ID.")) %>%
	mutate(Location = paste(seqnames, ":", start, "-", end, ":(", strand, ")", sep = "")) %>%
	dplyr::select(ID = gene_id,
				  Name = gene_name,
				  logFC, AveExpr, t, P.Value, 
				  FDR = adj.P.Val,
				  Location) %>%
	arrange(P.Value) %>%
	as_tibble()
```

```{r thresh}
thresh <- 0.2
```

- After analysis, `r sum(results$FDR < 0.05)` genes were found to be differentially expressed using an FDR of 0.05.
- Using a less stringent FDR threshold of `r percent(thresh)` gave weak-to-moderate support for `r sum(results$FDR < thresh)` genes as potentially DE.

```{r top10Table}
results %>%
	dplyr::filter(FDR < thresh) %>%
	pander(split.tables = Inf, 
		   caption = paste0("Top ", 
		   				nrow(.), 
		   				" most highly-ranked genes when sorting by p-value. ", "
		   				This represents an FDR of up to", percent(thresh), ".",
		   				"logFC > 0 indicates up regulation in diabetic samples relative to control"))
```


```{r plotMD, fig.cap=paste0("*Average expression against logFC. Confidently significant genes (FDR < 0.05) are indicated in red, whilst others are labelled to an FDR of ", percent(thresh), "*"), fig.width=8}
results %>%
	mutate(Sig = FDR < 0.05) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = Name),
					colour = "black",
					data = . %>% filter(FDR < thresh)) +
	scale_colour_manual(values = c(rgb(0.5,0.5,0.5,0.5), "red")) +
	theme(legend.position = "none")
```

```{r volcanoPlot, fig.cap = paste0("*Volcano plot with confidently significant genes shown in red. Genes to an FDR of", percent(thresh), "are labelled.*"), fig.height=7}
results %>%
	mutate(Sig = FDR < 0.05) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = Name),
					colour = "black",
					data = . %>% filter(FDR < thresh)) +
	scale_colour_manual(values = c(rgb(0.5,0.5,0.5,0.5), "red")) +
	theme(legend.position = "none")
```

```{r gnLabeller}
gnLabeller <- geneGR$gene_name %>%
	structure(names = geneGR$gene_id) %>%
	as_labeller()
```


```{r plotCPM, fig.cap=paste0("*Expression values for genes to an FDR of ", percent(thresh), ". Genes are shown in order of significance.*"), fig.width=8}
cpm(voomData, log = TRUE) %>%
	.[dplyr::filter(results, FDR < thresh)$ID,] %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(Rat, logCPM, -ID) %>%
	mutate(ID = factor(ID, levels = unique(ID))) %>%
	left_join(targets) %>%
	ggplot(aes(Group, logCPM, colour = Group)) +
	geom_jitter(width = 0.1) +
	facet_wrap(~ID, labeller = gnLabeller) +
	theme(legend.position = "none")
```


### Gene Set Enrichment

```{r human2Rat}
human2Rat <- read_tsv("Human2Rat.tsv") %>%
	dplyr::rename(ID = `Gene stable ID`,
				  Name = `Gene name`,
				  Entrez = `NCBI gene ID`)
idRankedByT <- results$t %>%
	set_names(results$ID) %>%
	sort()
```

```{r nperm}
np <- 1e6
```


Given the relatively small number of DE genes, the most suitable enrichment analysis was chosen to be GSEA as this works on a ranked list with no requirement for formal consideration of genes as differentially expressed.
This approach walks through a ranked list and checks to see if genes with a given gene-set are clustered at either end of the ranked list.
If no correlation between the gene-set and the experiment is evident, they should be evenly spread throughout the list.
Otherwise they could reasonably be expected to show directional bias.
As this is a fairly broad view analysis, gene sets were chosen as defined for human EntrezGene identifiers, then these were mapped to Ensembl rat identifiers.
All gene-sets are described at the [Broad Institute](http://software.broadinstitute.org/gsea/msigdb/collection_details.jsp)

Four gene sets were chosen:

- Hallmark gene-sets (k = 50): These are a small number of well defined, or high-level gene sets
- C2 Curated gene-sets (k = 4762): Gene sets curated from various sources such as online pathway databases, the biomedical literature, and knowledge of domain experts. 
- C5 GO gene-sets (k = 5917): Gene sets that contain genes annotated by the same GO term
- C7 Immunologic Signature gene-sets (k = 4872): Gene sets that represent cell states and perturbations within the immune system. The signatures were generated by manual curation of published studies in human and mouse immunology.

#### Hallmark Gene Sets

```{r loadHallmark}
hallmark <- gmtPathways("gmt/h.all.v6.2.entrez.gmt") %>%
	lapply(function(x){
		human2Rat %>%
			dplyr::filter(Entrez %in% x, 
						  ID %in% rownames(geneDGE)) %>%
			.[["ID"]]
	})
```


```{r testHallmark}
hallmarkRes <- fgsea(hallmark, idRankedByT, nperm=np) %>%
	as_tibble() %>%
	dplyr::rename(FDR = padj) %>%
	mutate(adjP = p.adjust(pval, "bonferroni")) %>%
	dplyr::select(pathway, pval, FDR, adjP, everything()) %>%
	arrange(pval)
```

```{r showHallmark}
hallmarkRes %>%
	dplyr::filter(adjP < 0.01) %>%
	mutate(sizeAtMax = vapply(leadingEdge, length, integer(1)),
		   DE = vapply(leadingEdge, function(x){
		   	results %>%
		   		dplyr::slice(1:100) %>%
		   		dplyr::filter(ID %in% x) %>%
		   		.$Name %>%
		   		paste(collapse = ",") 
		   }, character(1))) %>%
	dplyr::select(Pathway = pathway, NES, `$p$` = pval, `$p_{adj}$` = adjP, size, sizeAtMax, DE) %>%
	pander(split.tables = Inf,
		   caption = paste("Enriched Hallmark gene-sets using a Bonferroni-adjusted p-value < 0.01 for stringency.",
		   				"The normalised enrichment score indicates which end of the ranked list appeared to be enriched, with negative values indicating the down regulated genes.",
		   				"The number of genes in the gene-set are also indicated, as are the number of genes at the point of maximum enrichment.",
		   				"Any genes in the geneset which appeared in the most highly ranked 100 DE genes are shown to give a broad picture of which genes may be playing a role."),
		   style = "rmarkdown",
		   justify = "lrrrrrl")
```

```{r plotHallmark_1, fig.height=4, fig.cap = "*Enrichment score for the most highly ranked hallmark geneset*"}
tempSet <- "HALLMARK_INTERFERON_GAMMA_RESPONSE"
plotEnrichment(hallmark[[tempSet]], idRankedByT) +
	scale_x_continuous(breaks = seq(0, length(idRankedByT), by = 2e3)) +
	labs(title = tempSet)
```

#### C2 Curated Gene Sets

```{r loadC2}
c2 <- gmtPathways("gmt/c2.all.v6.2.entrez.gmt") %>%
	mclapply(function(x){
		human2Rat %>%
			dplyr::filter(Entrez %in% x, 
						  ID %in% rownames(geneDGE)) %>%
			.[["ID"]]
	}, mc.cores = 12)
```

As this was a larger database of gene-sets, only those with between 15 & 500 genes were tested.
To obtain greater statistical precision the number of permutations was increased to $1\times10^6$. 

```{r testC2}
c2Res <- fgsea(c2, idRankedByT, nperm=np, minSize = 15, maxSize = 500) %>%
	as_tibble() %>%
	dplyr::rename(FDR = padj) %>%
	mutate(adjP = p.adjust(pval, "bonferroni")) %>%
	dplyr::select(pathway, pval, FDR, adjP, everything()) %>%
	arrange(pval)
```

```{r showC2}
c2Res %>%
	dplyr::filter(adjP < 0.01) %>%
	mutate(sizeAtMax = vapply(leadingEdge, length, integer(1)),
		   DE = vapply(leadingEdge, function(x){
		   	results %>%
		   		dplyr::slice(1:100) %>%
		   		dplyr::filter(ID %in% x) %>%
		   		.$Name %>%
		   		paste(collapse = ",") 
		   }, character(1))) %>%
	dplyr::select(Pathway = pathway, NES, `$p$` = pval, `$p_{adj}$` = adjP, size, sizeAtMax, DE) %>%
	pander(split.tables = Inf,
		   caption = paste("Enriched C2 gene-sets using a Bonferroni-adjusted p-value < 0.01 for stringency.",
		   				"The normalised enrichment score indicates which end of the ranked list appeared to be enriched, with negative values indicating the down regulated genes.",
		   				"The number of genes in the gene-set are also indicated, as are the number of genes at the point of maximum enrichment.",
		   				"Any genes in the geneset which appeared in the most highly ranked 100 DE genes are shown to give a broad picture of which genes may be playing a role."),
		   style = "rmarkdown",
		   justify = "lrrrrrl")
```

```{r plotC2_1, fig.height=4, fig.cap = "*Enrichment score for the most highly ranked C2 geneset* As this peak was relatively far from the significantly DE genes, this is not a partiularly compelling result"}
tempSet <- c2Res$pathway[1]
plotEnrichment(c2[[tempSet]], idRankedByT) +
	scale_x_continuous(breaks = seq(0, length(idRankedByT), by = 2e3)) +
	labs(title = tempSet)
```

#### C5 GO Gene Sets

```{r loadC5}
c5 <- gmtPathways("gmt/c5.all.v6.2.entrez.gmt") %>%
	mclapply(function(x){
		human2Rat %>%
			dplyr::filter(Entrez %in% x, 
						  ID %in% rownames(geneDGE)) %>%
			.[["ID"]]
	}, mc.cores = 12)
```


```{r testC5}
c5Res <- fgsea(c5, idRankedByT, nperm=np, minSize = 15, maxSize = 500) %>%
	as_tibble() %>%
	dplyr::rename(FDR = padj) %>%
	mutate(adjP = p.adjust(pval, "bonferroni")) %>%
	dplyr::select(pathway, pval, FDR, adjP, everything()) %>%
	arrange(pval)
```

```{r showC5}
c5Res %>%
	dplyr::filter(adjP < 0.01) %>%
	mutate(sizeAtMax = vapply(leadingEdge, length, integer(1)),
		   DE = vapply(leadingEdge, function(x){
		   	results %>%
		   		dplyr::slice(1:100) %>%
		   		dplyr::filter(ID %in% x) %>%
		   		.$Name %>%
		   		paste(collapse = ",") 
		   }, character(1))) %>%
	dplyr::select(Pathway = pathway, NES, `$p$` = pval, `$p_{adj}$` = adjP, size, sizeAtMax, DE) %>%
	pander(split.tables = Inf,
		   caption = paste("Enriched C5 gene-sets using a Bonferroni-adjusted p-value < 0.01 for stringency.",
		   				"The normalised enrichment score indicates which end of the ranked list appeared to be enriched, with negative values indicating the down regulated genes.",
		   				"The number of genes in the gene-set are also indicated, as are the number of genes at the point of maximum enrichment.",
		   				"Any genes in the geneset which appeared in the most highly ranked 100 DE genes are shown to give a broad picture of which genes may be playing a role."),
		   style = "rmarkdown",
		   justify = "lrrrrrl")
```

```{r plotC5_1, fig.height=4, fig.cap = "*Enrichment score for the most highly ranked C2 geneset* As this peak was relatively far from the significantly DE genes, this is not a partiularly compelling result"}
tempSet <- c5Res$pathway[1]
plotEnrichment(c5[[tempSet]], idRankedByT) +
	scale_x_continuous(breaks = seq(0, length(idRankedByT), by = 2e3)) +
	labs(title = tempSet)
```


#### C7 Immunologic Signatures

```{r loadC7}
c7 <- gmtPathways("gmt/c7.all.v6.2.entrez.gmt") %>%
	mclapply(function(x){
		human2Rat %>%
			dplyr::filter(Entrez %in% x, 
						  ID %in% rownames(geneDGE)) %>%
			.[["ID"]]
	}, mc.cores = 12)
```


```{r testC7}
c7Res <- fgsea(c7, idRankedByT, nperm=np, minSize = 15, maxSize = 500) %>%
	as_tibble() %>%
	dplyr::rename(FDR = padj) %>%
	mutate(adjP = p.adjust(pval, "bonferroni")) %>%
	dplyr::select(pathway, pval, FDR, adjP, everything()) %>%
	arrange(pval)
```

```{r showC7}
c7Res %>%
	dplyr::filter(adjP < 0.01) %>%
	mutate(sizeAtMax = vapply(leadingEdge, length, integer(1)),
		   DE = vapply(leadingEdge, function(x){
		   	results %>%
		   		dplyr::slice(1:100) %>%
		   		dplyr::filter(ID %in% x) %>%
		   		.$Name %>%
		   		paste(collapse = ",") 
		   }, character(1))) %>%
	dplyr::select(Pathway = pathway, NES, `$p$` = pval, `$p_{adj}$` = adjP, size, sizeAtMax, DE) %>%
	pander(split.tables = Inf,
		   caption = paste("Enriched C7 gene-sets using a Bonferroni-adjusted p-value < 0.01 for stringency.",
		   				"The normalised enrichment score indicates which end of the ranked list appeared to be enriched, with negative values indicating the down regulated genes.",
		   				"The number of genes in the gene-set are also indicated, as are the number of genes at the point of maximum enrichment.",
		   				"Any genes in the geneset which appeared in the most highly ranked 100 DE genes are shown to give a broad picture of which genes may be playing a role."),
		   style = "rmarkdown",
		   justify = "lrrrrrl")
```

```{r plotC7_1, fig.height=4, fig.cap = "*Enrichment score for the most highly ranked C2 geneset* As this peak was relatively far from the significantly DE genes, this is not a partiularly compelling result"}
tempSet <- c7Res$pathway[1]
plotEnrichment(c7[[tempSet]], idRankedByT) +
	scale_x_continuous(breaks = seq(0, length(idRankedByT), by = 2e3)) +
	labs(title = tempSet)
```


<!-- ## RUV Analysis -->

<!-- In order to try and minimise the impact of library size,an alternate analysis was performed using RUV-Seq (Removing Unwanted Variation). -->

# Session Info

```{r sessionInfo}
pander(sessionInfo())
```

