---
title: "Quick Differential Gene Analysis"
author: "Steve Pederson"
date: "22 May 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
					  warning = FALSE, message = FALSE)
```

## Setup

```{r}
library(limma)
library(edgeR)
library(tidyverse)
library(magrittr)
library(pander)
library(ngsReports)
library(scales)
library(GenomicRanges)
library(rtracklayer)
library(plyranges)
library(ggrepel)
theme_set(theme_bw())
```

```{r}
rawFqc <- list.files("../0_rawData/FastQC", pattern = "zip", full.names = TRUE) %>%
	getFastqcData()
```


## Data Description

- The initial files were single-end reads of 100bp
- Library sizes were between `r pander(comma(range(readTotals(rawFqc)$Total_Sequences)))`
- Low levels of adapter content were found and removed using `AdapterRemoval`
- Spikes in GC content at around 63% and 72% were noted as potentially indicative of incomplete rRNA removal
    - rRNA reads were not removed as they will be ignored during summarisation
- Reads were aligned to the Rnor6.0 genome sourced from Ensembl Release 96, using `STAR 2.7.0d` for visualisation and inspection
- Reads were psuedo-aligned to the transcriptome from Ensembl Release 96 for summarisation to the transcript-level using `kallisto 0.43.1`. 
    + `kallisto` is able to proportionally assign multi-mapping reads
    + Transcript-level counts were then summarised to the gene-level.


```{r, fig.cap="*GC content of each library showing uneven and 'spikey' patterns. Peaks on the RHS indicate potential incomplete rRNA removal, whilst the other indicate possible excessive PCR cycles and high levels of duplicated sequences.*"}
plotGcContent(rawFqc, plotType = "line", theoreticalGC = TRUE, theoreticalType = "Transcriptome", species = "Rnorvegicus", pattern = "_F_.+")
```

```{r}
starLogs <- list.files(
	path = "../2_alignedData/log/",
	pattern = "final.out",
	full.names = TRUE) %>%
	importStarLogs()
```

```{r, fig.cap = "*STAR Alignment summary showing very low levels of unique alignments, which is again suggestive of incomplete rRNA removal. Uniquely mapping reads are now within the range usually seen after summarising to genes*"}
starLogs %>% 
	dplyr::select(Filename, contains("Number")) %>% 
	dplyr::select(Filename, contains("Mapped")) %>% 
	gather("Type", "Total", -Filename) %>% 
	mutate(Filename = str_remove(Filename, "_CB2Y.+"),
		   Type = str_remove(Type, "Number_Of_Reads_"),
		   Type = str_remove(Type, "_Reads_Number")) %>% 
	ggplot(aes(Filename, Total, fill = Type)) + 
	geom_bar(stat = "identity") +
	scale_y_continuous(label = comma) +
	labs(x = "Rat ID",
		 y = "Reads")
```

```{r, results='hide'}
counts <- list.files("../3_kallisto/", full.names = TRUE) %>% 
	catchKallisto()
```

```{r}
targets <- read_csv("targets.csv")
```


After summarisation the library sizes ranged between `r pander(comma(range(colSums(counts$counts))))`

- This is a disappointing result, but still manageable
- Low expressed genes will be aggressively removed as they will be unreliable given the small libraries
- Given the likelihood of high PCR duplicates, counts at the higher end will generally be more stable whilst those at the lower end will be less stable and removed.

## Analysis

```{r, cache=TRUE}
allGR <- import.gff("ftp://ftp.ensembl.org/pub/release-96/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.96.chr.gtf.gz")
geneGR <- allGR %>%
	subset(type == "gene") %>%
	plyranges::select(type, gene_id, gene_name, gene_biotype)
# transGR
transGR <- allGR %>%
	subset(type == "transcript") %>%
	plyranges::select(gene_id, gene_name, transcript_id, transcript_name, transcript_biotype)
```

```{r}
trans2Gene <- mcols(transGR) %>%
	as.data.frame() %>%
	dplyr::select(transcript_id, gene_id) %>%
	dplyr::filter(!is.na(transcript_id), !is.na(gene_id)) %>%
	as_tibble()
```

```{r}
geneDGE <- counts$counts %>%
	as.data.frame() %>%
	rownames_to_column("transcript_id") %>%
	as_tibble() %>%
	set_colnames(basename(colnames(.))) %>%
	set_colnames(str_remove(colnames(.),"_CB2Y.+")) %>%
	mutate(transcript_id = str_remove(transcript_id, "\\.[0-9]+")) %>%
	gather(Sample, Count, -transcript_id) %>%
	left_join(trans2Gene) %>%
	group_by(Sample, gene_id) %>%
	summarise(Count = sum(Count)) %>%
	spread(Sample, Count) %>% 
	dplyr::filter(grepl("ENSRNOG", gene_id)) %>%
	as.data.frame() %>% 
	column_to_rownames("gene_id") %>% 
	.[rowSums(cpm(.) > 2) >= 3,] %>%
	DGEList() %>%
	calcNormFactors()
geneDGE$samples %<>%
	mutate(Rat = rownames(.)) %>%
	left_join(targets) %>%
	set_rownames(.$Rat)
geneDGE$genes <- geneGR %>%
	magrittr::set_names(mcols(.)$gene_id) %>%
	.[rownames(geneDGE)]
```

```{r}
voomData <- voomWithQualityWeights(geneDGE)
```


```{r}
pca <- geneDGE %>%
	cpm(log = TRUE) %>%
	t() %>%
	prcomp()
summary(pca)
```

```{r}
plotly::ggplotly(pca$x %>%
	cbind(voomData$targets[rownames(.),]) %>%
	ggplot(aes(PC1, PC2, colour = Group, label = Rat)) +
	geom_point(aes(size = sample.weights)) +
	labs(x = paste0("PC1 (", percent(summary(pca)$importance["Proportion of Variance", "PC1"]), ")"),
		 y = paste0("PC2 (", percent(summary(pca)$importance["Proportion of Variance", "PC2"]), ")")) +
	theme(legend.position = "none"))
```

```{r}
results <- voomData %>%
	lmFit(design = model.matrix(~Group, data = .$targets)) %>%
	eBayes() %>%
	topTable(n = Inf) %>%
	set_colnames(str_remove(colnames(.),"ID.")) %>%
	mutate(Location = paste(seqnames, ":", start, "-", end, ":(", strand, ")", sep = "")) %>%
	dplyr::select(ID = gene_id,
				  Name = gene_name,
				  logFC, AveExpr, t, P.Value, 
				  FDR = adj.P.Val,
				  Location) %>%
	arrange(P.Value)
```

```{r}
head(results, 10) %>%
	pander(split.tables = Inf, 
		   caption = "Top ranked genes")
```


```{r}
thresh <- 0.15
results %>%
	mutate(Sig = FDR < 0.05) %>%
	ggplot(aes(AveExpr, logFC, colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = Name),
					data = . %>% filter(FDR < thresh)) +
	scale_colour_manual(values = c(rgb(0.5,0.5,0.5,0.5), "red")) +
	theme(legend.position = "none")
```

```{r}
results %>%
	mutate(Sig = FDR < 0.05) %>%
	ggplot(aes(logFC, -log10(P.Value), colour = Sig)) +
	geom_point() +
	geom_text_repel(aes(label = Name),
					data = . %>% filter(FDR < thresh)) +
	scale_colour_manual(values = c(rgb(0.5,0.5,0.5,0.5), "red")) +
	theme(legend.position = "none")
```

```{r}
gnLabeller <- as_labeller(structure(geneGR$gene_name, names = geneGR$gene_id))
```


```{r, fig.cap=paste("Genes to an FDR of", percent(thresh))}
cpm(voomData, log = TRUE) %>%
	.[dplyr::filter(results, FDR < thresh)$ID,] %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(Rat, logCPM, -ID) %>%
	left_join(targets) %>%
	ggplot(aes(Group, logCPM, colour = Group)) +
	geom_jitter(width = 0.1) +
	facet_wrap(~ID, labeller = gnLabeller)
```


```{r, fig.cap = "Genes up to an FDR of 20\\%"}
cpm(voomData, log = TRUE) %>%
	.[dplyr::filter(results, FDR > thresh, FDR < 0.2)$ID,] %>%
	as.data.frame() %>%
	rownames_to_column("ID") %>%
	gather(Rat, logCPM, -ID) %>%
	left_join(targets) %>%
	ggplot(aes(Group, logCPM, colour = Group)) +
	geom_jitter(width = 0.1) +
	facet_wrap(~ID, labeller = gnLabeller)
```

# Session Info

```{r}
pander(sessionInfo())
```

